File: samples/echo-java-enclave/src/main/java/com/r3/sgx/enclave/jecho/EchoEnclave.java
package com.r3.sgx.enclave.jecho;

import com.r3.conclave.common.internal.Cursor;
import com.r3.conclave.common.internal.SgxReportData;
import com.r3.sgx.core.enclave.EnclaveApi;
import com.r3.sgx.core.enclave.Enclavelet;

import java.nio.ByteBuffer;

public class EchoEnclave extends Enclavelet {
    @Override
    public Cursor<SgxReportData, ByteBuffer> createReportData(EnclaveApi api) {
        Cursor<SgxReportData, ByteBuffer> report = Cursor.allocate(SgxReportData.INSTANCE);
        ByteBuffer buffer = report.getBuffer();
        buffer.put(new byte[buffer.capacity()]);
        return report;
    }

    @Override
    public EchoHandler createHandler(EnclaveApi api) {
        return new EchoHandler();
    }
}

File: samples/echo-enclave/src/main/kotlin/com/r3/sgx/enclave/echo/EchoEnclave.kt
package com.r3.sgx.enclave.echo

import com.r3.conclave.common.internal.ByteCursor
import com.r3.conclave.common.internal.Cursor
import com.r3.conclave.common.internal.SgxReportData
import com.r3.sgx.core.common.Handler
import com.r3.sgx.core.common.Sender
import com.r3.sgx.core.enclave.EnclaveApi
import com.r3.sgx.core.enclave.Enclavelet
import java.nio.ByteBuffer
import java.util.function.Consumer

internal class EchoHandler : Handler<Sender> {
    override fun onReceive(connection: Sender, input: ByteBuffer) {
        connection.send(input.remaining(), Consumer { buffer ->
            buffer.put(input)
        })
    }

    override fun connect(upstream: Sender) = upstream
}

/**
 * An enclave that simply echoes the input to the enclavelet
 */
class EchoEnclave : Enclavelet() {
    override fun createReportData(api: EnclaveApi): ByteCursor<SgxReportData> {
        val report = Cursor.allocate(SgxReportData)
        val buffer = report.getBuffer()
        buffer.put(ByteArray(buffer.capacity()) {0})
        return report
    }

    override fun createHandler(api: EnclaveApi): Handler<*> {
        return EchoHandler()
    }
}
